src/scripts/main.js

```diff
@@ -1,3 +1,331 @@
+        if (game.matrix[col][row] === 0) {
+          continue;
+        }
+
+        checkSameTail(col, row, col - 1, row);
+      }
+    }
+  } else if (move === 'down') {
+    for (let col = 2; col >= 0; col--) {
+      for (let row = 0; row < 4; row++) {
+        if (game.matrix[col][row] === 0) {
+          continue;
+        }
+
+        checkSameTail(col, row, col + 1, row);
+      }
+    }
+  } else if (move === 'left') {
+    for (let col = 0; col < 4; col++) {
+      for (let row = 1; row < 4; row++) {
+        if (game.matrix[col][row] === 0) {
+          continue;
+        }
+
+        checkSameTail(col, row, col, row - 1);
+      }
+    }
+  } else if (move === 'right') {
+    for (let col = 0; col < 4; col++) {
+      for (let row = 2; row >= 0; row--) {
+        if (game.matrix[col][row] === 0) {
+          continue;
+        }
+
+        checkSameTail(col, row, col, row + 1);
+      }
+    }
+  }
+}
+
+function isGameLose() {
+  for (let i = 0; i < 4; i++) {
+    for (let j = 0; j < 4; j++) {
+      const current = game.matrix[i][j];
+
+      if (current === 0) {
+        return;
+      }
+
+      if (i !== game.matrix.length - 1 && current === game.matrix[i + 1][j]) {
+        return;
+      }
+
+      if (j !== game.matrix[i].length - 1
+        && current === game.matrix[i][j + 1]) {
+        return;
+      }
+    }
+  }
+
+  renderMessage('message-lose', 'You lose! Restart the game?');
+
+  window.removeEventListener('keydown', handlerKey);
+}
+
+function isGameWin() {
+  for (let i = 0; i < 4; i++) {
+    for (let j = 0; j < 4; j++) {
+      const current = game.matrix[i][j];
+
+      if (current === 2048) {
+        renderMessage('message-win', 'Winner! Congrats! You did it!');
+
+        window.removeEventListener('keydown', handlerKey);
+
+        return true;
+      }
+    }
+  }
+}
+
+function renderMessage(classMessage, message) {
+  messageContainer.innerHTML = `
+    <p class="message ${classMessage}">
+      ${message}
+    </p>
+  `;
+}
+
+function startGame() {
+  clearData();
+  addRandomCell();
+  addRandomCell();
+
+  render();
+
+  window.addEventListener('keydown', handlerKey);
+}
+
+function handlerKey(e) {
+  switch (e.key) {
+    case 'w':
+    case 'ArrowUp':
+      searchEmptyTails('up');
+      searchSameTails('up');
+      searchEmptyTails('up');
+
+      break;
+    case 's':
+    case 'ArrowDown':
+      searchEmptyTails('down');
+      searchSameTails('down');
+      searchEmptyTails('down');
+
+      break;
+    case 'a':
+    case 'ArrowLeft':
+      searchEmptyTails('left');
+      searchSameTails('left');
+      searchEmptyTails('left');
+
+      break;
+    case 'd':
+    case 'ArrowRight':
+      searchEmptyTails('right');
+      searchSameTails('right');
+      searchEmptyTails('right');
+
+      break;
+  }
+
+  render();
+
+  if (game.moved) {
+    setTimeout(() => {
+      gameScore.textContent = game.score;
+
+      addRandomCell();
+
+      render();
+    }, 240);
```

Comment:
I think it isn't a good idea to use delay here because I can make more than one move before the cell will appear